Chapter 20 web api rest (#18)
LaunchCodeEducation/csharp-web-dev-curriculum: content/web-api-rest/reading/web-api/_index.md

#### Universal Representation

It is necessary to adopt a universal representation because web APIs and client applications may be written in two different programming languages. Your web API may be written in C#/ASP.NET but the client application may be written using JavaScript and React. 

While there are many languages and frameworks available in web development, they all support the creation and parsing of JSON. JSON is a standard in web development because it is simple to process in any language, compatible with HTTP, and seamlessly represents the structure of data objects.

##### JSON

JSON is currently a universal representation of data accepted by client applications. This means our web API packages data 
requested by a client application as JSON. The web API also transfers this JSON in its communication with a client application.

<!-- TODO: Link will need to be changed from localhost once we have a permanent endpoint -->
Let's revisit the last two steps from our [web API work flow](http://localhost:1313/web-api-rest/web-api/#transference-of-data)

> 5. The controller transforms the object into a JSON representation.
> 6. The controller responds to the client with the JSON representation.

Reviewer: ```suggestion
1. The controller transforms the object into a JSON representation.
1. The controller responds to the client with the JSON representation.
```

{{% notice green "Tip" "rocket" %}}
[XML](https://developer.mozilla.org/en-US/docs/Web/XML/XML_introduction) is another popular data format. It is now used less commonly than 
JSON for web API-to-client communications.
{{% /notice %}}

> In the next section, we will discuss exactly how a client application makes a request and how a web API responds.

Reviewer: Is this referring to the section below? Might reword:
```suggestion
Let us discuss exactly how a client application makes a request and how a web API responds.
```

{{% notice green "Tip" "rocket" %}}
We will refer to web APIs as APIs going forward, since the web prefix is implied.
{{% /notice %}}

Here's a refresher on [the basics of HTTP](https://education.launchcode.org/intro-to-professional-web-dev/chapters/http/index.html):

- Is a stateless request/response protocol.
- Requests and responses may include HTTP bodies.
- Responses always contain a three digit HTTP status code.
- Requests and responses always include HTTP headers.

We call HTTP a *stateless protocol*. State can be a complex concept that refers to a number of things. We'll explore some aspects of it in more depth on the next page. 

In the context of HTTP, think of **state** as information about application data that is transferred via HTTP bodies, HTTP status codes, and HTTP headers.

#### Bodies

An HTTP body can contain a large number of different media types, 

> known as [MIME types](https://developer.mozilla.org/en-US/docs/Web/HTTP/Basics_of_HTTP/MIME_types/Common_types). A MIME type is associated with the HTTP header `Content-Type`. This header instructs the recipient of the HTTP request/response on what MIME type the HTTP body contains. We've seen a `Content-Type:text/html` HTTP header before. Here's a simple example of it:
> 

Reviewer: ```suggestion
known as [MIME types](https://developer.mozilla.org/en-US/docs/Web/HTTP/Basics_of_HTTP/MIME_types/Common_types). A MIME type is associated with the HTTP header `Content-Type`. This header instructs the recipient of the HTTP request/response on what MIME type the HTTP body contains. We've seen a `Content-Type:text/html` HTTP header before. 

{{% notice blue "Example" "rocket" %}}
```

> An HTTP body can contain a large number of different media types, 
> known as [MIME types](https://developer.mozilla.org/en-US/docs/Web/HTTP/Basics_of_HTTP/MIME_types/Common_types). A MIME type is associated with the HTTP header `Content-Type`. This header instructs the recipient of the HTTP request/response on what MIME type the HTTP body contains. We've seen a `Content-Type:text/html` HTTP header before. Here's a simple example of it:
> 
> ```html {linenos=table}
> <!DOCTYPE html>
> <html>
>    <head>
>       <title>My Web Page</title>
>       content
>    </head>
>    <body>
>       content
>    </body>
> </html>
> ```
> 
> This is the header for HTML documents and is used throughout the web. APIs send representations of data in the format of JSON requiring the header `Content-Type` to be `application/json`. This allows us to pass the state of the data as the HTTP body.
> 
> ```json {linenos=table}
> {
>    "title": "An Astronaut's Guide to Life on Earth",
>    "author": "Chris Hadfield",
>    "ISBN": 9780316253017,
>    "year_published": 2013,
>    "subject": ["Hadfield, Chris", "Astronauts", "Biography"],
>    "available": true
> }
> ```
> 
> The HTTP body may include JSON that represents the data being passed between an API and client application. Remember, not all requests/responses include HTTP bodies.

Reviewer: ```suggestion
An HTTP body can contain a large number of different media types, 
known as [MIME types](https://developer.mozilla.org/en-US/docs/Web/HTTP/Basics_of_HTTP/MIME_types/Common_types). A MIME type is associated with the HTTP header `Content-Type`. This header instructs the recipient of the HTTP request/response on what MIME type the HTTP body contains. We've seen a `Content-Type:text/html` HTTP header before. 

{{% notice green "Example" "rocket" %}}
```html {linenos=table}
<!DOCTYPE html>
<html>
   <head>
      <title>My Web Page</title>
      content
   </head>
   <body>
      content
   </body>
</html>
```

This is the header for HTML documents and is used throughout the web. APIs send representations of data in the format of JSON requiring the header `Content-Type` to be `application/json`. This allows us to pass the state of the data as the HTTP body.

```json {linenos=table}
{
   "title": "An Astronaut's Guide to Life on Earth",
   "author": "Chris Hadfield",
   "ISBN": 9780316253017,
   "year_published": 2013,
   "subject": ["Hadfield, Chris", "Astronauts", "Biography"],
   "available": true
}
```

The HTTP body may include JSON that represents the data being passed between an API and client application. Remember, not all requests/responses include HTTP bodies.
{{% notice %}}
```

> 

Reviewer: ```suggestion
{{% notice %}}

```

Reviewer: Put example in blue example box.
